### PR TITLE
Reorganization of the GHA files

### PR DESCRIPTION
**Pull request by TotoGaz**
The integrated tests can be run as part of the workflow when the label `ci: run integrated tests` is associated to the PR.

A bundle of the new reference files and of the logs is sent to the cloud and made available for download to the developer. (_For the sake of merging, this step deactivated in this PR._).

For the moment, integrated tests are only requested to be run on 2 procs. This can be changed easily.

Important note: while the integrated tests can be _run_, they are not _passing_:
```
    Status     : TestCase                                      : Directory                                                         : Elapsed : Resources : TestStep  
    ---------- : --------------------------------------------- : ----------------------------------------------------------------- : ------- : --------- : ----------
    FAIL RUN   : Sneddon_embeddedFrac_benchmark_01             : embeddedFractures/Sneddon_embeddedFrac_benchmark_01               : 0:00:03 : 0:00:03   : geos      
    ---------- : --------------------------------------------- : ----------------------------------------------------------------- : ------- : --------- : ----------
    FAIL CHECK : PoroElastic_permeableFault_smoke_01           : poroElasticCoupling/PoroElastic_permeableFault_smoke_01           : 0:00:04 : 0:00:04   : restartcheck
    FAIL CHECK : PoroElastic_impermeableFault_smoke_01         : poroElasticCoupling/PoroElastic_impermeableFault_smoke_01         : 0:00:04 : 0:00:04   : restartcheck
    FAIL CHECK : grav_seg_c1ppu_hyst_01                        : compositionalMultiphaseFlow/grav_seg_c1ppu_hyst_01                : 0:00:05 : 0:00:05   : restartcheck
    FAIL CHECK : ThermoPoroElastic_consolidation_sequential_02 : thermoPoromechanics/ThermoPoroElastic_consolidation_sequential_02 : 0:00:16 : 0:00:32   : restartcheck
    FAIL CHECK : grav_seg_c1ppu_hyst_02                        : compositionalMultiphaseFlow/grav_seg_c1ppu_hyst_02                : 0:00:03 : 0:00:06   : restartcheck
    ---------- : --------------------------------------------- : ----------------------------------------------------------------- : ------- : --------- : ----------
```

Also:
- The GHA `yaml` files now only manage input data and job dependencies.
In the end, the job informations are later provided to `ci_build_and_test_in_container.sh` directly.
Script `ci_build_and_test.sh`, which was previously calling `ci_build_and_test_in_container.sh` could be removed.
- The number of environment variables in the GHA workflow is reduced to nearly the maximum.
 (`LA` variables remain because I was afraid of conflicts. This can be done in a follow-up PR if wanted.)
 The scripts now use `getopt` instead.
- No duplicated call to the scripts for all the builds. One does all the cases with an decent number of `if`.
- Initial job `is_pull_request_a_draft` provides all the output requested from the PR (`assignees`, `draft`, ...).
It uses a single rest call and uses the GHA output pattern.
- No more final job to find if everything went OK; "_gate_" jobs allow to enter conditional jobs instead of using `if` in `yaml` which make the job optional (?). 
- The `GEOS_TPL_TAG` is not defined in `.github/workflows/ci_tests.yml` anymore. It's defined in and read from `.devcontainer/devcontainer.json` to avoid duplication. Documentation is updated accordingly.
- Extracting the data (`geos` executable bundle or integrated tests results) was simplified and do not require `docker cp` anymore.
The bundling process does not need to have access to out-of-scope information (like the commit or the PR number).
- The credentials for `sccache` is simplified a bit: no need for a specific mount point and copy of secrets.
- `codespaces` can now write in the `GEOS-DEV` submodules of the `GEOS` repository (in particular `integratedTests`).
- Fixes brittle assumptions when building the ats run scripts (in https://github.com/GEOS-DEV/integratedTests/pull/58)

Relates to issue https://github.com/GEOS-DEV/GEOS/issues/2800

### CODE REVIEW

**Review by CusiniM — APPROVED** [review]
@bmhan12, @rrsettgast if you are fine with this I would just go ahead and merge it.

File: scripts/ci_build_and_test_in_container.sh [hunk 1]
```diff
@@ -19,49 +24,131 @@ function or_die () {
...
+  --cmake-build-type ...
+      One of Debug, Release, RelWithDebInfo and MinSizeRel. Forwarded to CMAKE_BUILD_TYPE.
+  --data-basename output.tar.gz
+      If some data needs to be extracted from the build, the argument will define the tarball. Has to be a `tar.gz`.
+  --exchange-dir /path/to/exchange
+      Folder to share data with outside of the container.
+  --host-config host-config/my_config.cmake
+      The host-config. Path is relative to the root of the repository.
+  --install-dir-basename GEOS-e42ffc1
+      Geos installtion basename.
```
Comment (bmhan12) [inline scripts/ci_build_and_test_in_container.sh]:
```suggestion
      GEOS installation basename.
```

File: .github/workflows/ci_tests.yml [hunk 2]
```diff
@@ -6,65 +6,97 @@ concurrency:
...
-        ready_label="ready to be merged"
-        echo "Is PR Ready to be Merged? (Check for $ready_label label)"
-        echo "Labels found are: $labels"
-        if [[ $labels == *"$ready_label"* ]]; then "true" ; else "false"; fi
+        echo 'Labels are ${{ needs.is_pull_request_a_draft.outputs.LABELS }}'
+        ${{ contains( fromJSON( needs.is_pull_request_a_draft.outputs.LABELS ), 'ci: ready to be merged') }}
 
-  check_submodules:
-    needs: [check_pull_request_is_not_a_draft]
+  # Validates that the pr is still poiting to the HEAD of the main branch of the submodules repositories.
```
Comment (bmhan12) [inline .github/workflows/ci_tests.yml]:
```suggestion
  # Validates that the PR is still pointing to the HEAD of the main branch of the submodules repositories.
```

File: .github/workflows/ci_tests.yml [hunk 3]
```diff
@@ -75,84 +107,72 @@ jobs:
...
+      BUILD_AND_TEST_CLI_ARGS: ${{ matrix.BUILD_AND_TEST_ARGS }}
+      CMAKE_BUILD_TYPE: Release
+      DOCKER_IMAGE_TAG: ${{ needs.is_pull_request_a_draft.outputs.DOCKER_IMAGE_TAG }}
+      DOCKER_REPOSITORY: geosx/ubuntu20.04-gcc9
+      RUNS_ON: ubuntu-22.04
+      USE_SCCACHE: false
 
-  linux_builds:
+  # Matrix containing all the CPU build.
+  # Those are quite fase and can efficiently benefit from the `sccache' tool to make them even faster.
```
Comment (bmhan12) [inline .github/workflows/ci_tests.yml]:
```suggestion
  # Those are quite fast and can efficiently benefit from the `sccache' tool to make them even faster.
```

File: scripts/ci_build_and_test_in_container.sh [hunk 4]
```diff
@@ -75,47 +162,82 @@ fi
...
 fi
 
-if [[ ${USE_SCCACHE} ]]; then
+if [[ "${RUN_INTEGRATED_TESTS}" = true ]]; then
+  # We split the process in two steps. First installing the environment, then running the tests.
+  or_die ninja ats_environment
+  # The tests are not run using ninja (`ninja --verbose ats_run`) because it swallows the output while all the simulations are running.
+  # We directly use the script instead...
+  integratedTests/geos_ats.sh --failIfTestsFail
+  # Even (and even moreover) if the integrated testa fail, we want to pack the results for further investigations.
```
Comment (bmhan12) [inline scripts/ci_build_and_test_in_container.sh]:
```suggestion
  # Even (and even moreover) if the integrated tests fail, we want to pack the results for further investigations.
```

File: scripts/ci_build_and_test_in_container.sh [hunk 5]
```diff
@@ -19,49 +24,131 @@ function or_die () {
...
-
-if [[ "$*" == *--use-sccache* ]]; then
-  USE_SCCACHE=1
-fi
-
-if [[ ${USE_SCCACHE} ]]; then
-  mkdir -p ${HOME}/.config/sccache
-  cat <<EOT >> ${HOME}/.config/sccache/config
+if [[ ! -z "${SCCACHE_CREDS}" ]]; then
+  # The creadential json file is available at the root of the geos repository.
```
Comment (bmhan12) [inline scripts/ci_build_and_test_in_container.sh]:
```suggestion
  # The credential json file is available at the root of the geos repository.
```

File: scripts/ci_build_and_test_in_container.sh (hunk 4 above)
Comment (bmhan12) [inline scripts/ci_build_and_test_in_container.sh]:
```suggestion
  # Even (and even moreover) if the integrated tests fail, we want to pack the results for further investigations.
```

File: .github/workflows/ci_tests.yml [hunk 6]
```diff
@@ -6,65 +6,97 @@ concurrency:
...
     - name: Check that the PR is not a draft (cancel rest of jobs otherwise)
+      id: extract_pr_info
       run: |
-        draft_status=$(curl -H "Accept: application/vnd.github+json" \
-           https://api.github.com/repos/${{ github.repository }}/pulls/${{ github.event.number }} \
-           | jq '.draft')
-        echo "Is PR a draft? (Check PR's draft status )"
-        echo $draft_status
-        if [[ $draft_status == false ]]; then "true" ; else "false"; fi
+        # We do not rely on the `github.event.pull_request.labels` information since it's cached at the job.
```
Comment (bmhan12) [inline .github/workflows/ci_tests.yml:27]: 👍

**Review by bmhan12 — APPROVED** [review]
> No more final job to find if everything went OK; "gate" jobs allow to enter conditional jobs instead of using if in yaml which make the job optional (?).
 
Awesome! I just want to make sure: in the case where CUDA jobs are not run ('ci: ready to be merged' is not set), does the workflow show an expected failure status?
 
I found that without the convoluted "check_that_all_jobs_succeeded" job, a workflow will pass and the PR is mergeable without having ran the CUDA tests.

**Comment by TotoGaz** [comment]
> Awesome! I just want to make sure: in the case where CUDA jobs are not run ('ci: ready to be merged' is not set), does the workflow show an expected failure status?
> 
> I found that without the convoluted "check_that_all_jobs_succeeded" job, a workflow will pass and the PR is mergeable without having ran the CUDA tests.

Hi @bmhan12 and thanks for the review.
Yes, the workflow is tagged as "failed" if any of the `are_..._requested` jobs do not find their corresponding label. I've been checking that especially.
I think that if we use the `if:` in the `yaml`, then the jobs is _not_ done but considered _optional_...
I didn't find a way to fix this using GHA's `yaml` feature
Maybe it's possible, in the `yaml`, to do some `if not` and `exit 1`, but that duplicate the condition...

**Comment by TotoGaz** [comment]
Hi @bmhan12 I've just merged the PR. I did notice that the `check_that_all_jobs_succeeded` was `required` and I was not able to modify this (most likely it's in some GitHub admin pane I can't access). So I've put it back for the sake of merging. We can manage this in a follow-up PR where we'll manage the integrated tests as well.

**Pull request merged**